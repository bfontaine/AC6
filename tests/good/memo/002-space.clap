def square (n) =
  if n = 0 then 
    Cons [{ Hd <- Nil, Tl <- Nil }]
  else 
    val l = square (n - 1) in
    l . case {
    | Nil => Nil ** This case is impossible
    | Cons [{ Hd -> h, Tl -> l2 }] => 
      Cons [{ Hd <- Cons [{ Hd <- n, Tl <- h }], Tl <- l }]
    }

def sum (l) = 
  l . case {
  | Nil => 0
  | Cons [{ Hd -> h, Tl -> l }] => h + sum l
  }

def ssum (l) = 
  l . case {
  | Nil => 0
  | Cons [{ Hd -> l, Tl -> ls }] => sum l + ssum ls
  }

val bench = 
  val s = ssum (square (300)) in
  OK
    
