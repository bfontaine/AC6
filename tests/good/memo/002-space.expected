:- square = <code>
:- sum = <code>
:- ssum = <code>
:- bench = OK
